src/mlpack/methods/ann/layer/memory_head_impl.hpp

```diff
@@ -0,0 +1,482 @@
+#include "../visitor/gradient_visitor.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <typename InputDataType, typename OutputDataType>
+MemoryHead<InputDataType, OutputDataType>::MemoryHead() :
+    memoryHistory(dummyMemoryHistory),
+    dMem(dummyDMem)
+{
+
+}
+
+template <typename InputDataType, typename OutputDataType>
+MemoryHead<InputDataType, OutputDataType>::MemoryHead(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t memSize,
+    const size_t shiftRange,
+    const std::list<arma::mat>& memoryHistory,
+    arma::mat& dMem) :
+    inSize(inSize),
+    outSize(outSize),
+    memSize(memSize),
+    shiftSize(shiftRange),
+    forwardStep(0),
+    memoryHistory(memoryHistory),
+    dMem(dMem),
+    deterministic(false)
+{
+  // Build linear for kT + bT + gT + sT + gammaT
+  inputLinear = new Linear<>(inSize, (memSize) + (1) + (1) +
+      (2 * shiftSize + 1) + (1));
+
+  // kT non linearity.
+  kTNonLinear = new TanHLayer<>();
+
+  network.push_back(inputLinear);
+  network.push_back(kTNonLinear);
+
+  allZeros = arma::zeros<arma::mat>(outSize, 1);
+
+  prevWeights.push_back(std::move(arma::mat(allZeros.memptr(),
+      allZeros.n_rows, allZeros.n_cols, false, true)));
+
+  weightsBackwardIterator = prevWeights.end();
+
+  prevError = arma::zeros<arma::mat>((memSize) + (1) + (1) +
+      (2 * shiftSize + 1) + (1), 1);
+
+  bWdash = lWDash.end();
+  bGammaT = lGammaT.end();
+  bWTilde = lWTilde.end();
+  bShiftMatrix = lShiftMatrix.end();
+  bWg = lWg.end();
+  bSt = lSt.end();
+  bGt = lGt.end();
+  bWe = lWe.end();
+  bWc = lWc.end();
+  bBt = lBt.end();
+  bCosineT = lConsineT.end();
+  bMemoryHistory = memoryHistory.end();
+}
+
+template <typename InputDataType, typename OutputDataType>
+MemoryHead<InputDataType, OutputDataType>::~MemoryHead()
+{
+  boost::apply_visitor(deleteVisitor, inputLinear);
+  boost::apply_visitor(deleteVisitor, kTNonLinear);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void MemoryHead<InputDataType, OutputDataType>::Forward(arma::Mat<eT>&& input,
+                                                        arma::Mat<eT>&& output)
+{
+  const arma::mat& memory = memoryHistory.back();
+
+  // Pass the input through linear layer.
+  boost::apply_visitor(ForwardVisitor(std::move(input), std::move(
+      boost::apply_visitor(outputParameterVisitor, inputLinear))),
+      inputLinear);
+
+  arma::mat& lOutput = boost::apply_visitor(outputParameterVisitor,
+    inputLinear);
+
+  // Build kT with non linearity.
+  boost::apply_visitor(ForwardVisitor(std::move(lOutput.submat(0, 0,
+      memSize - 1, 0)), std::move(boost::apply_visitor(outputParameterVisitor,
+      kTNonLinear))), kTNonLinear);
+  const arma::mat& kT = boost::apply_visitor(outputParameterVisitor,
+      kTNonLinear);
+
+  // Build bT with non linearity
+  lBt.push_back(bTNonLinear.Fn(lOutput(memSize, 0)));
```

Comment:
Do you think instead of using a vector to hold the states, we could use ``arma::cube``? I think this would improve the overall readability.

```c++
lBt.push_back(bTNonLinear.Fn(lOutput(memSize, 0)));
const double& bT = lBt.back();
```

would become

```c++
gT.slice(step) = gTNonLinear.Fn(lOutput(memSize + 1, 0);
```